### DISCUSSION TITLE
Environment variable with hyphen/dash symbol not working

### DISCUSSION THREAD

**Discussion by asyba** [Kong/insomnia#3763] [Ideas]
I created an environment variable with symbol "-" and was not able to access it.

![image](https://user-images.githubusercontent.com/1466530/111809655-fa9c4180-88b3-11eb-8182-49226a7ad635.png)
![image](https://user-images.githubusercontent.com/1466530/111809683-012ab900-88b4-11eb-93a5-9e2a38988acb.png)
![image](https://user-images.githubusercontent.com/1466530/111809739-11429880-88b4-11eb-9fb7-dc5fe45ed93d.png)

I can't create another variable name, this is imported from third-party Open API specification.

Context
OS: Ubuntu 18.04.3 LTS
Insomnia v2021.1.1

**Comment by dimitropoulos** [comment]
my understanding is that for these environment variables, the rules of javascript variable naming apply.  dashes are not permitted in variable names in javascript (the parser will interpret them as a subtraction symbol) and thus are not permitted in environment variable names.

By the way, environment variables (as in, in shell or just in general) are not permitted to have dashes either.  https://pubs.opengroup.org/onlinepubs/009695299/basedefs/xbd_chap08.html

Could you possibly share what the third-party OpenAPI document is so we could take a look and see if there's another way to accomplish this?  I'm perplexed why this would have a hyphen in the first place.

**Comment by asyba** [comment]
> my understanding is that for these environment variables, the rules of javascript variable naming apply. dashes are not permitted in variable names in javascript (the parser will interpret them as a subtraction symbol) and thus are not permitted in environment variable names.
> 
> By the way, environment variables (as in, in shell or just in general) are not permitted to have dashes either. https://pubs.opengroup.org/onlinepubs/009695299/basedefs/xbd_chap08.html
> 
> Could you possibly share what the third-party OpenAPI document is so we could take a look and see if there's another way to accomplish this? I'm perplexed why this would have a hyphen in the first place.

I cannot share it but for example look at this online validator:
https://apitools.dev/swagger-parser/online/


```
{
  "openapi": "3.0.0",
  "info": {
    "title": "dummy",
    "version": "1.0",
    "license": {
      "name": "dummy"
    },
    "description": "dummy"
  },
  "paths": {
    "/api/v1/dummy/{dummy-token}": {
      "delete": {
        "summary": "dummy",
        "operationId": "dummy",
        "responses": {
          "204": {
            "$ref": "..dummy1.json"
          },
          "400": {
            "$ref": "..dummy2.json"
          }
        },
        "description": "dummy",
        "parameters": [
          {
            "$ref": "..dummy11.json"
          },
          {
            "$ref": "..dummy22.json"
          }
        ]
      }
    }
  }
}
```
![image](https://user-images.githubusercontent.com/1466530/112167334-22024f80-8bcf-11eb-982a-a0be67ed4dd4.png)

**Comment by asyba** [comment]
If import that:
![image](https://user-images.githubusercontent.com/1466530/112170115-90e0a800-8bd1-11eb-841c-3907c0b6830f.png)

**Comment by dimitropoulos** [comment]
that's very helpful, thanks.  @nijikokun it's to you, then, to make a product decision on how we want to proceed here.  We can (internally) probably get away with doing some kind of trickery (replace `-` with `DASH` or something for our internal representations, I donno, but (of course) never present this to the user) to get around this problem.  Naturally, that introduces new things (like checking for collisions) but all the same, worth exploring if this is in fact valid, as it appears to be.

**Comment by nijikokun** [comment]
I'm wondering if this is a potential regression from the recent importers work or pre-existing. I know that @develohpanda has put in work around adding support for special characters. #2601

**Comment by develohpanda** [comment]
Not a regression, I don't _think_ importers were updated to use the new variable format but I will need to confirm. Will get back to you!

**Comment by asyba** [comment]
@develohpanda any updates?

**Comment by develohpanda** [comment]
@asyba apologies for the delay, this seems to be the offending code (although the example should read `I.e. "/foo/{bar}" => "/foo/{{ bar }}" `), so I confirm the OpenAPI3 importer needs to be updated. https://github.com/Kong/insomnia/blob/bdb4b4e6610f69e84a0ae7ea1296078c59bbe88e/packages/insomnia-importers/src/importers/openapi3.js#L288-L298

Notes for someone picking this up:

The following spec generates a URL with template tags as shown below, but it should prefix the environment variable `abc-def` correctly as shown.

```yaml
openapi: 3.0.2

info:
  title: Example
  version: 0.0.1

servers:
  - url: https://example.com/v1

paths:
  '/api/{abc-def}':
    get:
      responses:
        '200':
          description: Hello
```

```diff
- {{ base_url }}/api/{{ abc-def }}
+ {{ base_url }}/api/{{ _['abc-def'] }}
```

You could take inspiration from the template tag generation logic here https://github.com/Kong/insomnia/pull/2601/files#diff-dbf418a3cd4acfc045b064cb1064e5f8cd1b05fc40a86b0b74d0606265176315R62-R70, or you could force a tag to always be generated with bracket notation (eg. `{{ _['base_url'] }}` is still valid JS syntax). For more context about this syntax change, read through #2601.

**Comment by asyba** [comment]
so what is the status?

**Comment by deepanshuds** [comment]
I am facing the same issue when importing collection. I saw that I can update the syntax of referencing the variable to {{_['parameter_name']}} but that adds an additional step for me to import multiple collections.

**Comment by subnetmarco** [comment]
We will look into this.

**Comment by subnetmarco** [comment]
@ihexxa correct me if I am wrong, we are currently solving for importing environment variables that have a hyphen `-` by automatically converting it to an underscore `_` ?
